pkg/controller/garbagecollector/rate_limiter_helper.go

```diff
@@ -29,31 +29,32 @@ import (
 // RegisteredRateLimiter records the registered RateLimters to avoid
 // duplication.
 type RegisteredRateLimiter struct {
-	rateLimiters map[unversioned.GroupVersion]struct{}
-	lock         sync.RWMutex
+	rateLimiters map[unversioned.GroupVersion]*sync.Once
 }
 
 // NewRegisteredRateLimiter returns a new RegisteredRateLimiater.
-func NewRegisteredRateLimiter() *RegisteredRateLimiter {
-	return &RegisteredRateLimiter{
-		rateLimiters: make(map[unversioned.GroupVersion]struct{}),
+// TODO: NewRegisteredRateLimiter is not dynamic. We need to find a better way
+// when GC dynamically change the resources it monitors.
+func NewRegisteredRateLimiter(resources []unversioned.GroupVersionResource) *RegisteredRateLimiter {
+	rateLimiters := make(map[unversioned.GroupVersion]*sync.Once)
+	for _, resource := range resources {
+		gv := resource.GroupVersion()
+		if _, found := rateLimiters[gv]; !found {
+			rateLimiters[gv] = &sync.Once{}
+		}
 	}
+	return &RegisteredRateLimiter{rateLimiters: rateLimiters}
 }
 
 func (r *RegisteredRateLimiter) registerIfNotPresent(gv unversioned.GroupVersion, client *dynamic.Client, prefix string) {
-	r.lock.RLock()
-	_, ok := r.rateLimiters[gv]
-	r.lock.RUnlock()
-	if ok {
+	once, found := r.rateLimiters[gv]
+	if !found {
 		return
 	}
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	if _, ok := r.rateLimiters[gv]; !ok {
+	once.Do(func() {
 		if rateLimiter := client.GetRateLimiter(); rateLimiter != nil {
 			group := strings.Replace(gv.Group, ".", ":", -1)
 			metrics.RegisterMetricAndTrackRateLimiterUsage(fmt.Sprintf("%s_%s_%s", prefix, group, gv.Version), rateLimiter)
```

Comment:
> what if another controller has a client for the same resource? won't that register the same metric twice?

`prefix` is different so the metric's name will be different. Also, different controller has uses different clientPool, so the same ratelimiter won't get registered twice.

> or in the dynamic garbage collector scenario, what if the user adds a 3rd party resource, then deletes it, then adds it again?

I left a TODO on line 36 saying this approach is not dynamic.